lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1424 @@
+//===- InferResets.cpp - Infer resets and add async reset -------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InferResets pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/FieldRef.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-resets"
+
+using llvm::BumpPtrAllocator;
+using llvm::MapVector;
+using llvm::SmallDenseSet;
+using llvm::SmallSetVector;
+using mlir::InferTypeOpInterface;
+using mlir::WalkOrder;
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+/// An absolute instance path.
+using InstancePathRef = ArrayRef<InstanceOp>;
+using InstancePath = SmallVector<InstanceOp>;
+
+template <typename T>
+static T &operator<<(T &os, InstancePathRef path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "/" << inst.name() << ":" << inst.moduleName();
+  return os;
+}
+
+static StringRef getTail(InstancePathRef path) {
+  if (path.empty())
+    return "$root";
+  auto last = path.back();
+  return last.name();
+}
```

Comment:
Yeah something like that would be really nice! We'd probably want to create dedicated subclasses of `ArrayRef` and `SmallVector` like llvm does in ADT :thinking:. Maybe @youngar has a bunch of ideas?

Reply:
This seems like a different thing, but could very much make sense as an analysis-like class.  Structuring it that way in a private header would make sense and make it easier to reason about this pass.  It can be done as a follow-up patch.

Reply:
Yeah that would make sense. Parts of GCT use a similar construct. Would probably make sense to have this around somewhere. Will do as a follow-up.